x86: fix syscall optimization issue

tyrantlizard77 opened the pull request:
These needed "memory" clobbers otherwise the compiler would do
unnecessary optimizations for parameters passed in as pointer
values.

Fixes an issue where system calls that had some arguments passed in via
struct were receiving all zeroes or garbage for those arguments.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

lpereira commented:
While the fix seems harmless, I find the compiler behavior quite surprising.  Parameters to inline assembly blocks are supposed to be completely opaque to the compiler.

andrewboie commented:
I'm not 100% sure what its going on.

Without the patch, the implementation of _syscall_invoke10, the compiler was skipping the initialization of the args struct and the whole thing would be whatever junk was in the stack at the time, usually zeroes.
If I changed the args struct declared in _syscall_invoke10() to be volatile, it fixes it.
If I changed the arg6 parameter of _arch_syscall_invoke6() to volatile, that also fixed it.
I went this route because it is contained within the x86 implementation.

thiagomacieira commented on include/arch/x86/arch.h:
```diff
@@ -576,7 +577,7 @@ static inline u32_t _arch_syscall_invoke6(u32_t arg1, u32_t arg2, u32_t arg3,
 #endif
 			 : "S" (call_id), "a" (arg1), "d" (arg2),
 			   "c" (arg3), "b" (arg4), "D" (arg5),
-			   [arg6] "m" (arg6));
```
You could also try making this an in-out argument ([arg6] "+m" (arg6)), maybe with early clobber ("+&m").

Using volatile looks hackish and fragile. 

Another suggestion is:

```c
    register u32_t a6 asm("ebp") = arg6;
    __asm__ volatile("int $0x80")"
 			 : "=a" (ret)
 #ifdef CONFIG_X86_IAMCU
 			   , "=d" (arg2), "=c" (arg3)
  #endif
  			 : "S" (call_id), "a" (arg1), "d" (arg2),
  			   "c" (arg3), "b" (arg4), "D" (arg5));
```

andrewboie replied:
> You could also try making this an in-out argument ([arg6] "+m" (arg6)), maybe with early clobber ("+&m").

I tried various combinations of these; none of them fix the issue

andrewboie replied:
> register u32_t a6 asm("ebp") = arg6;

This doesn't work either

andrewboie commented:
I haven't been able to find any alternate solutions that work, or don't leak this hack to the common code.

To reproduce:
clone https://github.com/andrewboie/zephyr/tree/memprotect-full
build tests/kernel/pipe/pipe_api for qemu_x86
If this change is reverted, you will see failures "FATAL: syscall failed check: margs->arg9 & K_USER"

thiagomacieira commented:
The "register" idea came from glibc's source code. It isn't exactly what they use because they make all registers use that formula, not just ebp. But if that isn't working, then this smells like a GCC optimiser bug.

lpereira commented:
This also feels like a GCC bug to me.  Might be a good idea to see if this can be reproduced on a newer build, and report it.  In the meantime, just to be sure -- as I'm not really confident the `volatile` is the right way to fix this -- I'd avoid using EBP altogether and pass the 6th parameter as a struct element.

andrewboie commented:
@lpereira ok I'll look into this.
Code currently assumes that any given arch can pass in 6 args via registers, but I'll look into making this configurable per-arch.

andrewboie commented:
Reducing args to 5 will require a huge amount of work. But it's a moot point. This isn't special to arg6.

What I am seeing is that passing in pointers to memory for any argument, the compiler gets confused. I think the u32_t cast may have something to with it. For example, I was trying to add support for 64-bit return values, for APIs like k_uptime_get():

```
static inline u64_t _syscall_ret64_invoke0(u32_t call_id)
{
	u64_t ret;

	_arch_syscall_invoke1((u32_t)&ret, call_id);
	return ret;
}
```

I get the following warning, even though I'm passing the address of ret in as a pointer (which the handler verifies and fills in with a 64-bit value)

```
/projects/zephyr/tests/kernel/threads/lifecycle/thread_init/outdir/qemu_x86/include/generated/syscalls/kernel.h: In function ‘k_uptime_get’:
/projects/zephyr/include/syscall.h:251:9: warning: ‘ret’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  return ret;
         ^~~
/projects/zephyr/include/syscall.h:248:8: note: ‘ret’ was declared here
  u64_t ret;
```

The warning goes away if I make the ret parameter volatile. I think this is the same issue and it is unrelated to the use of EBP.

thiagomacieira commented:
"m" means input, which would mean ret's address is used for input, but not output. Then again, the *address* value was used, the memory pointed by that address was modified. "memory" clobber missing?

andrewboie commented:
> "m" means input, which would mean ret's address is used for input, but not output

That's not what "m" means as defined by GCC. 

"‘m’
A memory operand is allowed, with any kind of address that the machine supports in general. Note that the letter used for the general memory constraint can be re-defined by a back end using the TARGET_MEM_CONSTRAINT macro."

And as I have been trying to explain, this is not specific to the 6th argument. The code above has no "m" constraint anywhere. Above it is set to "=a".

andrewboie commented:
Memory clobber also fixes it. I'm gonna go with that, thanks for the suggestion.

Commit:
x86: fix syscall optimization issue

These needed "memory" clobbers otherwise the compiler would do
unnecessary optimizations for parameters passed in as pointer
values.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Updated to use memory clobbers, I believe this is the root-cause fix, thanks @thiagomacieira !

thiagomacieira commented:
> > "m" means input, which would mean ret's address is used for input, but not output

> That's not what "m" means as defined by GCC.

The point is that it doesn't have "+" or "=" prefix, so it's an input memory operand.

thiagomacieira reviewed (commented):
Looks good.
